### PR TITLE
xds: use the new cert-provider instances if present

### PR DESCRIPTION
**Pull request by sanjaypujare**
After importing the new protos (definitions) in PR #8490 this PR uses the new fields as per the gRFC [A29](https://github.com/grpc/proposal/blob/master/A29-xds-tls-security.md#xds-protocol). Specifically:

- use [`tls_certificate_provider_instance`](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L247) if present
- use [`ca_certificate_provider_instance`](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/transport_sockets/tls/v3/common.proto#L315) if present in either [`validation_context`](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L261) or [`combined_validation_context.default_validation_context`](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L188)

Various validations and translations take into account the above.

### CODE REVIEW

File: xds/src/main/java/io/grpc/xds/ClientXdsClient.java [hunk 1]
```diff
@@ -492,35 +488,33 @@ static void validateCommonTlsContext(
             "common-tls-context with tls_certificate_certificate_provider is not supported");
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/ClientXdsClient.java:488]: Why throw on `hasTlsCertificateCertificateProvider()` only when `certInstanceName == null`?

File: xds/src/main/java/io/grpc/xds/ClientXdsClient.java [hunk 2]
```diff
@@ -473,8 +469,8 @@ static void validateCommonTlsContext(
           "common-tls-context with validation_context_certificate_provider_instance is not"
               + " supported");
     }
-    String certInstanceName = null;
-    if (!commonTlsContext.hasTlsCertificateCertificateProviderInstance()) {
+    String certInstanceName = getIdentityCertInstanceName(commonTlsContext);
+    if (certInstanceName == null) {
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/ClientXdsClient.java:473]: The error messages `"common-tls-context with tls_certificates is not supported"`,  and `"common-tls-context with tls_certificate_sds_secret_configs is not supported"` seems misleading. Should mention `TlsCertificateProviderInstance` is unset.

File: xds/src/main/java/io/grpc/xds/ClientXdsClient.java [hunk 3]
```diff
@@ -547,13 +541,38 @@ static void validateCommonTlsContext(
...
-      if (certProviderInstances == null || !certProviderInstances.contains(rootCaInstanceName)) {
-        throw new ResourceInvalidException(
-            "ValidationContextProvider instance name '" + rootCaInstanceName
-                + "' not defined in the bootstrap file.");
+  }
+
+  private static String getIdentityCertInstanceName(CommonTlsContext commonTlsContext) {
+    if (commonTlsContext.hasTlsCertificateProviderInstance()) {
+      return commonTlsContext.getTlsCertificateProviderInstance().getInstanceName();
+    } else if (commonTlsContext.hasTlsCertificateCertificateProviderInstance()) {
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/ClientXdsClient.java:549]: This is deprecated, are we still supporting it?

Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/ClientXdsClient.java:549]: Yes, until Traffic Director is updated to make this change we will have to support the older/deprecated fields

File: xds/src/main/java/io/grpc/xds/ClientXdsClient.java (hunk 1 above)
Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/ClientXdsClient.java:488]: Good question! So `certInstanceName` is derived from the new field `tls_certificate_provider_instance` (or in its absence the old field `tls_certificate_certificate_provider_instance` which we will stop supporting as soon as the control plane makes the change). So if `certInstanceName == null` that means those fields are not set. And in such a situation if the control plane has populated the other fields that serve the same purpose (`tls_certificate_certificate_provider` or `tls_certificate_sds_secret_configs` or `tls_certificates`) then we have to generate this error.

However when `certInstanceName != null` then that means the field(s) we support is present and in that case even if those other fields are populated we don't complain. This way the control plane (TD) is able to send a payload that works for both gRPC and Envoy.

File: xds/src/main/java/io/grpc/xds/internal/certprovider/CertProviderSslContextProvider.java [hunk 4]
```diff
@@ -88,6 +90,50 @@ private static CertificateProviderInfo getCertProviderConfig(
     return certProviders != null ? certProviders.get(pluginInstanceName) : null;
   }
 
+  protected static CertificateProviderInstance getCertProviderInstance(
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/internal/certprovider/CertProviderSslContextProvider.java:93]: `@Nullable`?

File: xds/src/main/java/io/grpc/xds/internal/certprovider/CertProviderSslContextProvider.java [hunk 5]
```diff
@@ -88,6 +90,50 @@ private static CertificateProviderInfo getCertProviderConfig(
...
+    CertificateProviderInstance certInstance = null;
+    if (commonTlsContext.hasTlsCertificateProviderInstance()) {
+      return CommonTlsContextUtil.convert(commonTlsContext.getTlsCertificateProviderInstance());
+    } else if (commonTlsContext.hasTlsCertificateCertificateProviderInstance()) {
+      certInstance = commonTlsContext.getTlsCertificateCertificateProviderInstance();
+    }
+    return certInstance;
+  }
+
+  protected static CertificateValidationContext getStaticValidationContext(
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/internal/certprovider/CertProviderSslContextProvider.java:104]: `@Nullable`

File: xds/src/main/java/io/grpc/xds/internal/certprovider/CertProviderSslContextProvider.java [hunk 6]
```diff
@@ -88,6 +90,50 @@ private static CertificateProviderInfo getCertProviderConfig(
...
+      CommonTlsContext.CombinedCertificateValidationContext combinedValidationContext =
+          commonTlsContext.getCombinedValidationContext();
+      if (combinedValidationContext.hasDefaultValidationContext()) {
+        return combinedValidationContext.getDefaultValidationContext();
+      }
+    }
+    return null;
+  }
+
+  protected static CommonTlsContext.CertificateProviderInstance getRootCertProviderInstance(
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/internal/certprovider/CertProviderSslContextProvider.java:118]: `@Nullable`

File: xds/src/main/java/io/grpc/xds/ClientXdsClient.java [hunk 7]
```diff
@@ -492,35 +488,33 @@ static void validateCommonTlsContext(
             "common-tls-context with tls_certificate_certificate_provider is not supported");
       }
     } else {
-      certInstanceName = commonTlsContext.getTlsCertificateCertificateProviderInstance()
-          .getInstanceName();
-    }
-    if (certInstanceName != null) {
       if (certProviderInstances == null || !certProviderInstances.contains(certInstanceName)) {
```
[the remark this answers is not in the document]
Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/ClientXdsClient.java:491]: good suggestion. Done

File: xds/src/main/java/io/grpc/xds/ClientXdsClient.java (hunk 2 above)
Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/ClientXdsClient.java:473]: fixed.

**Review by dapengzhang0 — APPROVED** [review]

**Pull request merged**